### PR TITLE
[Diffusion][HunyuanImage3] Add request paged prefix caching

### PR DESCRIPTION
**Pull request by zwhzzz0821**
## Purpose

Add cross-request prefix caching for the HunyuanImage3 single DiT stage. Requests that share the same reference image and VAE random state but use different prompts can reuse prefix KV, reducing first-step computation; partial hits and CFG dual branches are supported.

The design has three layers: the model defines the caching semantics, the Scheduler manages logical blocks, and the Worker holds the physical KV pages and runs the model.

```mermaid
flowchart LR
    A["Model preprocessing<br/>token IDs + reference image identity/position"] --> B["Scheduler / native KVCacheManager<br/>block hash → hit lookup → allocate remaining blocks"]
    B -->|"block IDs + cached_prefix_len"| C["Worker<br/>install BlockTable and slot mapping"]
    C --> D["DiT first step<br/>compute only uncached queries, read full KV"]
    D --> E["After request succeeds<br/>publish stable blocks, release request references"]
    E --> B
```

1. **The model defines the reusable prefix.** The sequence is split into a stable text/reference-image prefix and a generated-image region that changes every step; only complete blocks inside the prefix are cached. Reference-image identity covers the actual image input, the VAE seed/generator state, and position/grid information — not just an image placeholder. Model semantics and LoRA ID/scale are isolated via namespaces.
2. **The Scheduler reuses upstream cache management.** The model writes the image identity and token intervals into the upstream `MultiModalFeatureSpec`; an upstream helper derives extra hash keys, which are chained with the token IDs and the previous block's hash to form the identity. Requests with different prompts can therefore still hit the common prefix before they diverge. The native `KVCacheManager` handles lookup, allocation, reference counting, and reclamation — no new allocator is added. The two CFG rows are looked up separately, executed at their common hit length, and rolled back as a whole if allocation fails.
3. **Worker/DiT skips already-hit computation.** The Worker installs the physical mapping from the block IDs and hit length issued by the Scheduler. At the first step the model trims the hit query-side inputs while preserving absolute positions, the full KV range, and the corresponding attention mask; the remaining queries can still attend to the cached prefix. Later denoise steps reuse the stable KV covering the dynamic region. VAE/ViT encoding is not skipped on KV hits.

Only successfully completed requests publish complete stable blocks; failed or cancelled requests publish nothing, and non-block-aligned boundary blocks and dynamic latents remain request-private.

Enabled via `diffusion_kv_mode=paged_scheduler` + `enable_prefix_caching=true`; off by default, with no content-hashing overhead when disabled. This PR does not cover combining AR KV import with prefix hits.

## Test Plan

Model: `tencent/HunyuanImage-3.0-Instruct`; 4×NVIDIA L20X, EP enabled, CFGP1, CFG=2.5, seed=42, 1280×720.

Both experiments reuse the two-reference-image [IT2I input](tests/assets/hunyuan_image3/it2i.jsonl) on a single DiT stage.

| Test | Scenario and checks |
| --- | --- |
| [E2E benchmark](tests/dfx/perf/tests/test_hunyuan_image3_prefix_caching.json) | Uses `vllm bench serve --omni`; TP4/SP1, 8 steps, concurrency 1, 2 warmups excluded and 8 measured identical requests per mode. Compares dense, paged no-cache, and paged prefix. Measures full-prefix reuse. |
| [Reference-image accuracy](tests/e2e/accuracy/test_hunyuan_image3_prefix_cache_accuracy.py) | 50 steps; TP4/SP1 and TP2/strict-Ulysses-SP2. A different prompt warms the cache; the original IT2I prompt then exercises partial and repeated hits. Verifies hits into both image spans, aligned CFG boundaries, actual query trimming, and repeat determinism. |
| Unit regression | Cache identity, allocation/publication/cleanup, mixed-boundary admission, configuration validation, and native attention backend selection. |

Accuracy compares generated outputs against the repository's [official Instruct output](tests/assets/hunyuan_image3/hunyuan_image_instruct_ref.png), using the existing IT2I thresholds—not against input images or uncached outputs. The golden comes from AR→DiT; this is a single-DiT output-quality check, not proof of identical conditioning.

**vLLM Version:** `0.29.0`; PyTorch `2.13.0`, torchvision `0.28.0`, Transformers `5.14.1`, Diffusers `0.38.0`.

**vLLM-Omni Commit:** Performance and TP4/SP1 accuracy: `ef07cc7de` plus the recorded workspace patch. TP2/SP2 accuracy: `994f80025`.

## Test Result

Local validation: **358 KV/Scheduler unit tests passed**, all **24 measured benchmark requests succeeded**, and real-hit accuracy passed under **both TP4/SP1 and TP2/SP2**.

### Performance

8 steps, TP4/SP1; one measured round.

| Mode | Throughput (req/s) ↑ | Mean latency (s) ↓ | P99 (s) ↓ |
| --- | ---: | ---: | ---: |
| dense | 0.3341 | 2.9893 | 3.0265 |
| paged no-cache | 0.3353 | 2.9773 | 3.0023 |
| paged prefix | **0.3932** | **2.5393** | **2.5674** |

Prefix caching improves throughput by **17.69% over dense** and **17.24% over paged no-cache** on this repeated-request workload.

### Accuracy against official output

All scored outputs passed: **CLIP ≥90, SSIM ≥0.26, PSNR ≥12.5 dB**.

| Topology | Output | CLIP | SSIM | PSNR (dB) |
| --- | --- | ---: | ---: | ---: |
| TP4/SP1 | dense | 97.5957 | 0.8040 | 17.4521 |
| TP4/SP1 | paged no-cache | 97.6941 | 0.8010 | 17.5476 |
| TP4/SP1 | prefix partial hit | 97.7520 | 0.8100 | 17.5651 |
| TP4/SP1 | prefix repeated hit | 97.2923 | 0.8099 | 17.6957 |
| TP2/SP2 | dense | 97.7008 | 0.8202 | 17.9827 |
| TP2/SP2 | paged no-cache | 97.9755 | 0.8075 | 17.6424 |
| TP2/SP2 | prefix partial hit | 97.4446 | 0.8091 | 17.7346 |
| TP2/SP2 | prefix repeated hit | 97.3744 | 0.8101 | 17.7021 |

Both topologies verified the following in each CFG branch:

| Case | Cached tokens | First-step query tokens |
| --- | ---: | ---: |
| No cache | 0 | 14,613 |
| Different-prompt partial hit | 10,976 | 3,637 |
| Repeated-request hit | 11,008 | 3,605 |

The three repeated-hit outputs were byte-identical within each topology.

### Generated outputs

Official AR→DiT golden and single-DiT outputs from the TP2/strict-Ulysses-SP2 run; CFG=2.5, seed=42, 50 steps. All Omni panels use the original IT2I prompt. Images are resized only for display.

image

CC @hsliuustc0106 @Bounty-hunter @asukaqaq-s @Acerak01-fy

### CODE REVIEW

File: benchmarks/diffusion/diffusion_benchmark_serving.py [hunk 1]
```diff
@@ -639,9 +639,12 @@ def _resolve_image_paths(self, item: dict) -> list[str] | None:
             if isinstance(paths, str):
                 paths = [paths]
 
-            # Verify all paths exist
+            # Preserve existing cwd-relative paths; also accept paths relative
```
Comment (Gaohan123) [inline benchmarks/diffusion/diffusion_benchmark_serving.py:642]: Currently, vllm bench already supports image generation. Please add new functions to vllm bench, rather than here

File: tests/dfx/perf/fixtures/hunyuan_image3_prefix_params.json [hunk 2]
```diff
@@ -0,0 +1,74 @@
+[
+  {
```
Comment (Gaohan123) [inline tests/dfx/perf/fixtures/hunyuan_image3_prefix_params.json:2]: PTAL. @yenuo26

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py [hunk 3]
```diff
@@ -1312,6 +1327,69 @@ def _prepare_attention_mask_for_generation(
         model_kwargs["full_attn_spans"] = full_attn_spans
         return attention_mask
 
+    @staticmethod
+    def _slice_cached_prefix_inputs(
```
Comment (Gaohan123) [inline vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py:1331]: For this method, is it model specific logic or general logic which can be reused?

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 4]
```diff
@@ -708,14 +709,38 @@ def _execute_request_list(
                         f"metadata={len(diffusion_kv_metadata)}, requests={len(reqs)}"
                     )
                 paged_metadata = self._build_paged_attention_metadata(diffusion_kv_metadata)
+                cached_prefix_lens = {row.kv_start_pos for row in paged_metadata.prefill_rows}
+                if len(cached_prefix_lens) != 1:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py]: [P2] Mixed prefix-hit boundaries in one first-step wave fail every scheduled request.

With `max_num_seqs > 1`, `_can_schedule_waiting` batches on sampling params only, so a repeat request that hit 5856 tokens and a partial-hit request that stopped at 4096 (exactly the shared-reference workload this PR targets) can be reserved in the same `schedule()` wave and reach this check together. The `ValueError` then propagates to the engine, which converts it into `FINISHED_ERROR` for every `scheduled_request_ids` entry (diffusion_engine.py:598-612) — one mixed-hit burst terminally fails the whole wave instead of degrading to a slower execution.

Suggest deferring at admission instead: `reserve_request` already returns the wave's `cached_prefix_len` in its metadata, so the admission loop can compare it against the pending first-step boundary and hold differing-boundary requests in `waiting` for the next wave.

**Comment by hsliuustc0106** [comment]
Reviewed at `54c891cb`. Identity design and lifecycle look sound: the block-hash chain roots at a deterministic per-namespace digest (stable across manager instances, isolated from native `NONE_HASH`), reference identity covers actual image-tensor bytes plus the VAE random state, LoRA id/scale live in the namespace, `get_computed_blocks`' 3-tuple / `allocate_slots` / `touch`-then-`free_blocks(reversed(...))` / `cache_blocks` usage all match the pinned vLLM v0.28.0 contracts, and publish-before-free ordering keeps `publish_request` KeyError-safe. Unit coverage for hit/miss/partial/CFG-min-boundary/unpublished-failure is strong. Two P3s and two validation gaps:

- [P3] `_DiffusionConfigProjection` omits `enable_prefix_caching` (vllm_omni/config/omni_config.py:650) — the legacy engine-args path routes the stage YAML key into `OmniDiffusionConfig.from_kwargs`, but the typed projection used by the RFC #4021 stage-init cutover drops unknown keys silently in `from_kwargs`, and `diffusion_kv_mode` is already mirrored there. Adding the field alongside keeps the knob from silently no-op'ing after the cutover.
- [P3] Docs: `docs/design/feature/prefix_caching.md:44` still says prefix caching "is only supported on AutoRegressive stages with one kv-cache group" — this PR extends it to diffusion (`diffusion_kv_mode=paged_scheduler` + `enable_prefix_caching`, HunyuanImage3), so the feature page (and the diffusion support/compat matrix) should be updated in this PR or a fast-follow.
- Validation gap: only DCO ran; no test lane has executed the new unit/accuracy tests in CI (I can't add CI-trigger labels — please push a lane, e.g. `ready`).
- Validation gap: the GPU evidence ran SP1; the Ulysses first-step path changed its key/value slicing for trimmed queries (`hunyuan_image3_transformer.py:1125-1144`). It's mechanically consistent and the paged adapter packs joint text into the same paged row, but an SP>1 accuracy pass with real hits would close the only untested topology.

Except for the P2 batching hole (mixed hit boundaries in one wave fail all scheduled requests), this is approve-able to me.

**Comment by Bounty-hunter** [comment]
It seem that there is an accuracy in paged no cache base.

File: vllm_omni/diffusion/diffusion_engine.py [hunk 5]
```diff
@@ -872,13 +874,25 @@ def make_engine(
         engine.run_startup_warmup()
         return engine
 
+    def _prefix_cache_enabled(self) -> bool:
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/diffusion_engine.py:1058]: `enable_prefix_caching=true` currently becomes a silent no-op for paged models without a registered prefix-cache hook (and for unsupported diffusion modes). Please fail fast with an actionable config error.

File: docs/design/feature/prefix_caching.md [hunk 6]
```diff
@@ -167,6 +167,57 @@ When we pass our multimodal tensors to the language model component in the same
...
+denoise forward computes only the uncached query suffix while attending to both
+cached and new KV. Dynamic target-image KV is never published as a reusable prefix.
+
+Cache identity includes token IDs, reference-image content and VAE random state,
+plus model/layout and LoRA context. The same image with different prompts can
+reuse the common leading blocks; it does not imply that every image span or CFG
+branch is interchangeable. Disabling prefix caching skips cache-identity hashing;
+`dense_legacy` remains the default.
+
+The current scope is local DiT reuse, not AR-imported KV or missing-page-only
```
Comment (Bounty-hunter) [inline docs/design/feature/prefix_caching.md:200]: The SP>1 trimmed-query path is mechanically consistent, but the supplied real-hit evidence covers SP1; the existing SP4 regression is T2I/no-hit. Please add one SP>1 reference-prefix accuracy test with a non-zero hit, or temporarily reject/document that topology until it is validated.

**Review by Bounty-hunter — COMMENTED** [review]
LGTM — no blocking correctness or architectural issue found; the two items above are robustness/validation hardening.

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py (hunk 3 above)
Comment (zwhzzz0821) [reply vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py:1439]: This is currently model-specific. A general runner-side implementation needs more careful design to accommodate BAGEL’s packed context prefills and MiniMax-H3’s packed text/video/audio inputs, with their cacheability validated separately.
Hunyuan’s slicing depends on embeddings, attention masks, positional encodings, and timestep scatter indices constructed inside the pipeline. The runner currently receives the request, not these prepared tensors, so moving this logic requires separating input preparation from execution and defining a shared input contract.
I suggest keeping it model-specific in this PR and addressing that generalization in a follow-up.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py (hunk 4 above)
Comment (zwhzzz0821) [reply vllm_omni/diffusion/worker/diffusion_model_runner.py]: fixed

File: tests/dfx/perf/fixtures/hunyuan_image3_prefix_params.json (hunk 2 above)
Comment (zwhzzz0821) [reply tests/dfx/perf/fixtures/hunyuan_image3_prefix_params.json]: removed. Now the test only changes the existing it2i test deployment method, so these contents are no longer needed.

File: benchmarks/diffusion/diffusion_benchmark_serving.py (hunk 1 above)
Comment (zwhzzz0821) [reply benchmarks/diffusion/diffusion_benchmark_serving.py]: Updated to use vllm bench serve --omni with the custom_image dataset and /v1/images/edits.

File: vllm_omni/diffusion/diffusion_engine.py (hunk 5 above)
Comment (zwhzzz0821) [reply vllm_omni/diffusion/diffusion_engine.py:1058]: Currently, it directly throws an error when only prefix cache is enabled and page is not enabled.

File: docs/design/feature/prefix_caching.md (hunk 6 above)
Comment (zwhzzz0821) [reply docs/design/feature/prefix_caching.md:200]: The SP>1 trimmed-query path is mechanically consistent, but the supplied real-hit evidence covers SP1; the existing SP4 regression is T2I/no-hit. Please add one SP>1 reference-prefix accuracy test with a non-zero hit, or temporarily reject/document that topology until it is validated.

File: vllm_omni/diffusion/data.py [hunk 7]
```diff
@@ -1182,6 +1185,13 @@ def __post_init__(self):
         if not isinstance(self.diffusion_compile_dynamic, bool):
             raise TypeError(f"diffusion_compile_dynamic must be a bool, got {type(self.diffusion_compile_dynamic)!r}")
         self.diffusion_kv_mode = parse_diffusion_kv_cache_mode(self.diffusion_kv_mode)
+        if not isinstance(self.enable_prefix_caching, bool):
+            raise TypeError("enable_prefix_caching must be a bool")
```
Comment (Gaohan123) [inline vllm_omni/diffusion/data.py:1200]: [P1] Invalidate cached prefixes across sleep/wake or reject this combination

This validation still allows `enable_prefix_caching=True` together with `enable_sleep_mode=True`. After a successful request publishes its prefix, `sleep(level=1)` preserves only the `weights` allocations and discards the `kv_cache` contents. `wake_up()` refreshes the worker's block tables, but neither the diffusion sleep/wake RPC path nor the scheduler clears the native prefix-cache index.

Consequently, a subsequent request with the same prefix receives a nonzero `cached_prefix_len` and skips recomputing KV that no longer exists, potentially producing silently corrupted output.

Please invalidate the scheduler's cached prefixes when the KV pages are discarded, or reject prefix caching with sleep mode until that lifecycle is supported. A regression should cover warming the cache, sleeping and waking the engine, then repeating the request and verifying that discarded KV is recomputed.

Comment (zwhzzz0821) [reply vllm_omni/diffusion/data.py:1200]: The combination of sleep_mode and prefix_cache is now rejected. If later we need to let these two work together, it should also require only relatively few changes; simply disabling indexing during the preprocessing stage will suffice.

**Review by Gaohan123 — APPROVED** [review]
LGTM. Thanks

**Comment by Gaohan123** [comment]
Please resolve conflicts

**Comment by zwhzzz0821** [comment]
> Please resolve conflicts

done

**Comment by zwhzzz0821** [comment]
fixed the CI error. PTAL @Gaohan123

**Comment by zwhzzz0821** [comment]
This doc error looks not related to this pr.

**Comment by Gaohan123** [comment]
Please resolve conflicts
